### DISCUSSION TITLE
Mapbox with Codemagic

### DISCUSSION THREAD

**Discussion by IbrahimKhanLS** [codemagic-ci-cd/codemagic-docs#2775] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

Building IOS project for a specific flavor

### Expected results

It should successfully build the iOS SDK as I am using the same NETRC_MAPBOX_LOGIN, NETRC_MAPBOX_MACHINE and SDK_REGISTRY_TOKEN for other flavors and they are building fine for iOS using codemagic.

### Actual results

Error installing Mapbox-iOS-SDK
curl: (22) The requested URL returned error: 401

It works fine when building iOS SDK through my local project

**Comment by dam-ease** [comment]
Hello @IbrahimKhanLS 
This seems to be an issue with missing .netrc file. I found a similar issue [here](https://github.com/mapbox/mapbox-gl-native/issues/16581) in which adding the file as detailed [here](https://github.com/mapbox/mapbox-gl-native/issues/16581#issuecomment-784794472) solves. Can you please add the script below to add the file:

```
touch ~/.netrc
echo "machine api.mapbox.com" > ~/.netrc
echo "login mapbox" >> ~/.netrc
echo "password <INSERT API TOKEN>" >> ~/.netrc
```

If you run into any further issues, kindly share and also include the build ID. You can find it under your build details page's 'ID' section. ﻿Navigate to the 'Builds' section on the left-hand menu, select the build you want to view, and click the copy icon.

**Reply by IbrahimKhanLS** [reply]
Hey @dam-ease 
I am trying to make a build file using codemagic (CI/CD tool), and I have already added NETRC_MAPBOX_LOGIN, NETRC_MAPBOX_MACHINE and SDK_REGISTRY_TOKEN in the environment variable.

**Reply by dam-ease** [reply]
Hey @IbrahimKhanLS,
Yes, got that these have been added to env variables! Can you provide more information on your last message i.e. making a build file.  Have you added the script I shared earlier to your YAML file? Please share any updates you have on the issue? And if there is anything else I can assist you with, please don't hesitate to let me know.

**Reply by IbrahimKhanLS** [reply]
![Screenshot 2024-10-17 at 8 44 10 AM](https://github.com/user-attachments/assets/8df93270-0745-4f0f-8219-ca17a5a775e2)
Error shows unauthorized access to download Mapbox iOS sdk. I have added the mapbox details in the environment variable but not in the YAML file as I am using the default workflow. Secondly I have 3 other flavors that uses the same environment variable and source code including YAML file but they are build successfully.
@dam-ease

**Reply by IbrahimKhanLS** [reply]
@dam-ease the best part is manual upload works perfectly.

**Reply by dam-ease** [reply]
Please can you share the buildID's of the failed build with this error message as well as that of the two successful builds with me?
You can find it under your build details page's 'ID' section. ﻿Navigate to the 'Builds' section on the left-hand menu, select the build you want to view, and click the copy icon.

**Comment by IbrahimKhanLS** [comment]
@dam-ease 
Failed build id: 6710972913036e8871c33a92
Success build id: 670fafd73b7bb929db96fa6e
Success build id: 6706205d9526f7fce1cf299a

**Reply by dam-ease** [reply]
Hey @IbrahimKhanLS,
In the successful builds, I see that you are creating the `.netrc` file and adding the variables to it in the pre-build script, hence why the error doesn't show up in these, can you check this out?

**Reply by IbrahimKhanLS** [reply]
@dam-ease that is what my concern is. I am not running custom workflows based on different flavors. I am using default workflow from the codemagic UI and same source code for all flavors. Initially it was building fine and then suddenly it stopped building.
Is there any way I can manually restrict a specific flavor to run pre-build script?

**Reply by dam-ease** [reply]
Alright, instead of making use of just the default workflow, you could actually duplicate this workflow, creating several workflows for building different configurations of your app(same source code). To do this, simply navigate to App settings > Right sidebar > Workflow settings and click Duplicate workflow.

> Is there any way I can manually restrict a specific flavor to run pre-build script?

Yes, to do this you can make use of environment variables, this option enables you set environment variables based on the flavor being built. You can then modify your pre-build step to check for the specific flavor and only run the script for certain flavors. Here’s a sample snippet you can use in the pre-build script:
```
if [ "$BUILD_FLAVOR" == "specific_flavor" ]; then
  echo "Running pre-build script for $BUILD_FLAVOR..."
  # Add your pre-build commands here
else
  echo "Skipping pre-build script for $BUILD_FLAVOR..."
fi
```

**Reply by IbrahimKhanLS** [reply]
@dam-ease you want me to add this code in the codemagic.yaml file?

**Reply by dam-ease** [reply, marked as answer]
> default workflow from the codemagic UI 

Oh, I interpreted this as that you are making use of the Workflow editor now. I also see this looking through your buildID. If this is the case, you can add it to the section in the screenshot below:
<img width="858" alt="Screenshot 2024-10-23 at 09 24 41" src="https://github.com/user-attachments/assets/fb761c94-bb43-464e-95aa-1738f6daea95">
If not, then you can add it before your build script in your YAML file.

**Reply by IbrahimKhanLS** [reply]
@dam-ease thanks, adding command in the pre-build section worked.
